### ISSUE TITLE
Create UndefinedJinjaVariablesRule to ease upgrade to 2.0

### ISSUE THREAD

**Issue by potiuk** [apache/airflow#11144]
This issue is part of #8765

## Rule
Create `UndefinedJinjaVariablesRule` which corresponds to
> Prior to Airflow 2.0 Jinja Templates would permit the use of undefined variables. They would render as an
empty string, with no indication to the user an undefined variable was used. With this release, any template
rendering involving undefined variables will fail the task, as well as displaying an error in the UI when
rendering.

entry in UPDATING.md. Introduced in #11016 
This rule should check if any of the jinja templates use undefined variables. This might not be always possible, as this might involve the dynamic generation of some context variables, but likely we can at least print some warning or "double-check this" statements.

## How to guide
To implement a new rule, create a class that inherits from `airflow.upgrade.rules.base_rule.BaseRule`.
It will be auto-registered and used by `airflow upgrade-check` command. The custom rule class has to have `title`,
`description` properties and should implement `check` method which returns a list of error messages in case of
incompatibility.

For example:
https://github.com/apache/airflow/blob/ea36166961ca35fc51ddc262ec245590c3e236fb/airflow/upgrade/rules/conn_type_is_not_nullable.py#L25-L42

**Remember** to open the PR against `v1-10-test` branch.

**Comment by ashmeet13** [comment]
Hi, I am new to both Airflow and contributing to FOSS.
Saw this issue unpicked and I would like to pick this up.

I had one quick query - 
The base branch from which I create a new feature branch in my fork would be `v1-10-test`?

Thanks! Looking forward to contribute and learn more!

**Comment by potiuk** [comment]
Yep.

**Comment by ashmeet13** [comment]
I had a few doubts asking them here - 

Currently I was writing the following logic for `check` - 
For a given dag - 
1. It will gather all the tasks - `dag.tasks`
2. Iterate over the task list and for every task
        - Get Jinja2 environment from `task.get_template_env()`
        - Find Jinja2 variables for all `task.template_fields`
        - Verify that all the above gathered variables are present in `task.params`
3. Return a list of messages for every missing variable in `task.params`

Below is the script version

```
            for task in dag.tasks:
                env = task.get_template_env()
                params = set(task.params.keys())
                for attr_name in task.template_fields:
                    content = getattr(task, attr_name)
                    parsed = env.parse(content)
                    template_varaibles = meta.find_undeclared_variables(parsed) # Returns a set of variables present in the template
                    undefined_variables = template_varaibles-params # Set difference
```

Does this logic look like something we could follow?
And if yes then how can I write this logic to check for every DAG? 


I did have a look at `DagBag` but I do not know how should I write a test for it.
Sorry for the newbie questions - would be really helpful if you could guide on these points.

**Comment by ashb** [comment]
@ashmeet13 That check looks like a good start yes, but not quite there. I don't think your `params` is right. `TaskInstance.get_template_context()` is the parameters that templates are executed in, and for that you will need to create a TaskInstance (it can be in memory, no need to write it to the DB)

**Comment by ashb** [comment]
Also check our BaseOperator.render_template for all the possible locations/field types we render templates on. You'll need to replicate much of that logic to handle all the types (inside lists, dicts, etc. etc.)

**Comment by ashmeet13** [comment]
Got it - thank you so much. Will update on the progress or in case of any further doubts!

**Comment by ashmeet13** [comment]
Some progress that I have made. Thanks @ashb  for guiding towards `TaskInstance` for getting the complete context.
Also towards `BaseOperator.render_template`

With this I was able to improve the logic to extract undefined variables in a template to this - 

```
def check_rendered(rendered):
    if isinstance(rendered, six.string_types):
        return set(re.findall(r"{{(.*?)}}", rendered))

    elif isinstance(rendered, (tuple,list,set)):
        parsed_templates = set()
        for element in rendered:
            parsed_templates.union(check_rendered(element))
        return parsed_templates

    elif isinstance(rendered, dict):
        parsed_templates = set()
        for key, value in rendered.items():
            parsed_templates.union(check_rendered(value))
        return parsed_templates 



def check(dag):
    dag.template_undefined = jinja2.DebugUndefined
    for task in dag.tasks:
        jinja_env = task.get_template_env()
        task_instance = TaskInstance(task=task, execution_date=timezone.utcnow())
        template_context = task_instance.get_template_context()

        for attr_name in task.template_fields:
            content = getattr(task, attr_name)
            if content:
                rendered = task.render_template(content, template_context)
                undefined = check_rendered(rendered)
                for element in undefined:
                    print(element) 
```

Take into the consideration the following DAG -
```
dag = DAG(
    'tutorial',
    default_args=default_args,
    description='A simple tutorial DAG',
    schedule_interval=timedelta(days=1),
)

templated_command_string = """
{% for i in range(5) %}
    echo "{{ params.defined }}"
    echo "{{ execution_date.month }}"
    echo "{{ params.undefined }}"
    echo "{{ foo }}"
{% endfor %}
"""

t1 = BashOperator(
    task_id='templated_string',
    depends_on_past=False,
    bash_command=templated_command_string,
    params={'defined': 'test_val'},
    dag=dag,
)
```

The above check would print/return this -  
```
foo
no such element: dict object['undefined'] 
```


One blocker still that I have is how exactly will I be able to get the dags when the user runs `airflow upgrade-check`
Thanks again for the help!

**Comment by turbaszek** [comment]
Closed via #11241
